src/ir/opt/cfold.rs

```diff
@@ -244,6 +244,29 @@ pub fn fold_cache(node: &Term, cache: &mut TermCache<Term>) -> Term {
                     b.width() + w,
                 ))))
             }),
+            Op::Map(op) => match (get(0).as_array_opt(), get(1).as_array_opt()) {
+                (Some(a), Some(b)) => {
+                    // TODO: extend for n-ary arrays
+                    let mut res = a.clone();
+                    let mut merge = ArrayMerge::new(a.clone(), b.clone());
+                    for (i, va, vb) in merge.into_iter() {
+                        let r = fold_cache(
+                            &term![*op.clone(); leaf_term(Op::Const(va.clone())), leaf_term(Op::Const(vb.clone()))],
```

Comment:
As currently written 3 lines above, ArrayMerge owns the `a` and `b` data. So it seems like ArrayMerge could return `va` and `vb` by value (because it owns those values and doesn't need them any more)---yet they're cloned again here. That seems wasteful.

I'd suggest having ArrayMerge take `a` and `b` by shared reference, per my prior comment, and either

1. have it return `va` and `vb` as references, in which case they'd be cloned here, or
2. have ArrayMerge clone `va` and `vb` and return them as values, in which case no need to clone here.

All else equal, it would be best if ArrayMerge returned references, because that's the most generic thing: if the caller needs values (as here), it can clone; but if it only needs references there hasn't been any time or memory wasted on cloning.